execute/observation.go

```diff
@@ -333,49 +374,98 @@ func (p *Plugin) getMessagesObservation(
-	observationOptimizer := optimizers.NewObservationOptimizer(
-		lggr,
-		maxObservationLength,
-		p.ocrTypeCodec,
-	)
-	observation, err = observationOptimizer.TruncateObservation(observation)
-	if err != nil {
-		return exectypes.Observation{}, fmt.Errorf("unable to truncate observation: %w", err)
+		// Read messages for this report's sequence number range
+		msgs, err := p.readMessagesForReport(ctx, lggr, srcChain, report)
+		if err != nil {
+			lggr.Errorw("unable to read all messages for report",
+				"srcChain", srcChain,
+				"seqRange", report.SequenceNumberRange,
+				"merkleRoot", report.MerkleRoot,
+				"err", err,
+			)
+			continue
+		}
+
+		// Add the report to available reports
+		availableReports[srcChain] = append(availableReports[srcChain], report)
+		// Initialize data structures for this source chain if needed
+		initSourceChainMaps(srcChain, messageObs, msgHashes, tkData)
+
+		// Initialize msg hashes with empty messages and empty token data
+		// for all messages in the report
+		// We need to have it like that to be able to build the merkle tree and proofs in later stages
+		//nolint:lll // link
+		// https://github.com/smartcontractkit/chainlink-ccip/blob/5d360b7c90126631c51f3e84f26c0417dc3877b6/execute/report/roots.go#L14-L14
+		for _, msg := range msgs {
+			seqNum := msg.Header.SequenceNumber
+			messageObs[srcChain][seqNum] = createEmptyMessageWithIDAndSeqNum(msg)
+			tkData[srcChain][seqNum] = exectypes.NewMessageTokenData()
+			hash, err := p.msgHasher.Hash(ctx, msg)
+			if err != nil {
+				return exectypes.Observation{}, fmt.Errorf(
+					"unable to hash message srcChain %d, seqNum %d, msgId %v: %w",
+					srcChain, seqNum, msg.Header.MessageID, err,
+				)
+			}
+			msgHashes[srcChain][seqNum] = hash
+		}
+
+		observation.CommitReports = availableReports
+		observation.Hashes = msgHashes
+		observation.Messages = messageObs
+		observation.TokenData = tkData
+
+		// Process each message in the report and override the empty message and token data if everything fits within
+		// the size limits
+		for _, msg := range msgs {
+			// If msg is not already inflight, add it
+			if p.inflightMessageCache.IsInflight(srcChain, msg.Header.MessageID) {
+				continue
+			}
+			seqNum := msg.Header.SequenceNumber
+			messageObs[srcChain][seqNum] = msg
+			tkData[srcChain][seqNum] = p.observeTokenDataForMessage(ctx, lggr, msg)
+			observation.Messages = messageObs
+			observation.TokenData = tkData
+			totalMsgs++
+
+			if totalMsgs > lenientMaxMsgsPerObs ||
+				exceedsMaxEncodingSize(observation, p.ocrTypeCodec, lenientMaxObservationLength) {
+				// remove the last message and token data
+				observation.Messages[srcChain][seqNum] = createEmptyMessageWithIDAndSeqNum(msg)
+				observation.TokenData[srcChain][seqNum] = exectypes.NewMessageTokenData()
+				stop = true
+				break
+			}
+		}
+
+		if stop {
+			lggr.Infow("Stop processing messages, observation is too large")
```

Comment:
Please include some more context here, like `totalMsgs` and if possible the last observation size in case we are debugging issues.